hawkmoth/docstring.py

```diff
@@ -198,6 +198,35 @@ class TextDocstring(Docstring):
     _indent = 0
     _fmt = '\n'
 
+    def get_name(self):
+        """Figure out a name for the text comment based on the comment contents.
+
+        The name is the sub-string starting from the first alphanumeric
+        character in the comment to the next :, ., or newline.
+
+        This sensibly covers cases like reStructuredText hyperlink targets::
+
+            .. _Foo Bar:
+
+        and section titles::
+
+            Foo Bar
+            =======
+
+        and just first sentences::
+
+            Foo Bar. Blah.
+
+        Not perfect, but good enough.
```

Comment:
Hmmm, this looks simple, but suddenly a user documents two files with the same initial line and Sphinx starts complaining about duplicate reference tags.

Food for thought: anonymous symbols get a completely unguessable and non permanent tag right now. Here, a filename might be a good alternative, but for anonymous types we'd need file and line number, which is maybe too brittle and still not 'permanent'.

A common solution to these cases might be to generate throw away, unguessable and non permanent tag by default if we must or none if we don't, and instruct the user to specify a tag in those cases as needed. E.g.:

```c
/**
 * .. _tag:
 * Top level comment.
 */
```

In the anonymous types' case this would produce a link to the body of the block rather than the declaration line though I think.

Reply:
> Hmmm, this looks simple, but suddenly a user documents two files with the same initial line and Sphinx starts complaining about duplicate reference tags.

That's actually not true. It's not automatically generating any reference targets for Sphinx.

The user can add hyperlink targets themselves, and if they add it as the first thing in the comment, it also works for the purpose of selecting the comment. But then they're adding the targets, and it's their responsibility to handle any conflicts.

Reply:
Bah, sorry, this case has no directive of course, so there isn't an automatic reference tag being generated.

As for filtering the right comments, right now the file argument is mandatory, but even afterwards it can be used to disambiguate if need be. That still leaves comments in the same file starting the same way as a potential issue, right? Not sure we want to solve that though.

Reply:
> That still leaves comments in the same file starting the same way as a potential issue, right? Not sure we want to solve that though.

Even that's only a problem if the user wants to pick one or the other; I think with the current implementation it'll probably pick both.

I'm in no way claiming this naming system is ideal, but AFAICS everything else pretty much requires proper parsing of the comment for some structured identifier for the comment. For example requiring some `:meta id: name` field list and parsing using `extract_metadata()` from `sphinx.util.docstrings`. But that also requires stripping the comment indicators first and overall complicates matters.

I think a generic documentation comment would typically have a heading or a descriptive first sentence or a reference target, and that just could work as a name.

Reply:
Yup, I see your points. But if you don't mind, let it marinate a bit longer. This will be part of our API... Better think it through now than later.